### PR TITLE
[PKP-LIB][main] craft-oa BL Initial Commit

### PR DESCRIPTION
**Pull request by defstat**

### CODE REVIEW

File: pages/invitation/index.php [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+ * @ingroup pages_submission
+ *
+ * @brief Handle requests for the submission wizard.
+ *
+ */
+
+switch ($op) {
+    case 'decline':
+    case 'accept':
+        define('HANDLER_CLASS', 'PKP\pages\invitation\PKPInvitationHandler');
```
Comment (asmecher) [inline pages/invitation/index.php:19]: The use of `HANDLER_CLASS` is deprecated; the page index wrappers can just return the handler class directly. See e.g.: https://github.com/pkp/ojs/blob/main/pages/about/index.php

**Review by asmecher — CHANGES_REQUESTED** [review]
Thanks, @defstat! I've done a quick first review; have a look through it and I'll do a second round once you're satisfied.

File: classes/invitation/invitations/BaseInvitation.php [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+use Symfony\Component\Mailer\Exception\TransportException;
+
+abstract class BaseInvitation
+{
+    use SerializesModels;
+
+    /**
+     * The name of the connection the job should be sent to.
+     *
+     * @var string|null
```
[the remark this answers is not in the document]
Comment (defstat) [reply classes/invitation/invitations/BaseInvitation.php:42]: Fixed

File: classes/invitation/invitations/BaseInvitation.php [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+    public DateTime $expirationDate;
+
+    private Invitation $invitationModel;
+    /**
+     * eMail
+     */
+    protected string $email;
+
+    public function __construct(string $type, string $email, int $contextId, ?int $assocId)
```
[the remark this answers is not in the document]
Comment (defstat) [reply classes/invitation/invitations/BaseInvitation.php:63]: That's beautiful. Fixed..

File: classes/invitation/invitations/BaseInvitation.php [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    {
+        $this->invitationMarkStatus(InvitationStatus::INVITATION_STATUS_DECLINED);
+
+        return true;
+    }
+
+    abstract public function getInvitationMailable() : Mailable;
+    abstract public function preDispatchActions() : bool;
+
+    public function getAcceptInvitationURL() : string
```
[the remark this answers is not in the document]
Comment (defstat) [reply classes/invitation/invitations/BaseInvitation.php:134]: Fixed

File: pages/invitation/PKPInvitationHandler.php [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+use APP\core\Request;
+use APP\facades\Repo;
+use APP\handler\Handler;
+use APP\template\TemplateManager;
+use PKP\invitation\invitations\BaseInvitation;
+use ReflectionClass;
+
+class PKPInvitationHandler extends Handler
+{
+    public const INVITATION_REPLY_BASE = 'invitation';
```
[the remark this answers is not in the document]
Comment (defstat) [reply pages/invitation/PKPInvitationHandler.php:30]: Changed

File: pages/invitation/PKPInvitationHandler.php [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    public const INVITATION_REPLY_ACCEPT = 'accept';
```
[the remark this answers is not in the document]
Comment (defstat) [reply pages/invitation/PKPInvitationHandler.php:31]: Fixed

File: pages/invitation/index.php (hunk 1 above)
Comment (defstat) [reply pages/invitation/index.php:19]: Fixed.

File: classes/invitation/repositories/Invitation.php [hunk 7]
```diff
@@ -0,0 +1,93 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * @file classes/job/repositories/Job.php
+ *
+ * Copyright (c) 2014-2022 Simon Fraser University
```
[the remark this answers is not in the document]
Comment (defstat) [reply classes/invitation/repositories/Invitation.php:8]: Changed

File: classes/invitation/repositories/BaseRepository.php [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    }
+
+    public function perPage(int $perPage): self
+    {
+        $this->perPage = $perPage;
+
+        return $this;
+    }
+
+    // public function deleteJobs(string $queue = null, array $ids = []): int
```
[the remark this answers is not in the document]
Comment (defstat) [reply classes/invitation/repositories/BaseRepository.php:96]: Removed - I will check again

File: classes/invitation/models/Invitation.php [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+        'status' => 'int',
+        'context_id' => 'int',
+        'type' => 'string',
+        'invitation_email' => 'string',
+    ];
+
+    /**
+     * Instantiate the Job model
+     */
+    public function __construct(array $attributes = [])
```
[the remark this answers is not in the document]
Comment (defstat) [reply classes/invitation/models/Invitation.php:82]: Removed

File: classes/invitation/invitations/BaseInvitation.php [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+
+    public function __construct(string $type, string $email, int $contextId, ?int $assocId)
+    {
+        $this->type = $type;
+        $this->email = $email;
+        $this->contextId = $contextId;
+        $this->assocId = $assocId;
+
+        // This should be taken in another place or taken in a journal bases?
+        $this->expirationDate = Carbon::now()->addDays(3)->toDateTime();
```
[the remark this answers is not in the document]
Comment (defstat) [reply classes/invitation/invitations/BaseInvitation.php:71]: Changed that to this `$this->expirationDate = Carbon::now()->addDays(Config::getVar('invitations', 'default_expiration_days'))->toDateTime();` and added a new 

```
[invitations]
default_expiration_days = 3
```
in `config.TEMPLATE.inc.php`

File: classes/invitation/invitations/user/UserBoardPosition.php [hunk 11]
```diff
@@ -0,0 +1,72 @@
+<?php
+
+/**
+ * @file invitation/invitations/user/UserBoardPosition.php
```
[the remark this answers is not in the document]
Comment (defstat) [reply classes/invitation/invitations/user/UserBoardPosition.php:4]: Removed all subclasses from this PR.

File: classes/invitation/invitations/BaseInvitation.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (asmecher) [reply classes/invitation/invitations/BaseInvitation.php:71]: I'd suggest calling it just `expiration_days` (I don't think there's any way to change the default other than this setting) -- and then provide the `3` default as a 3rd parameter to `Config::getVar` for those who don't have the setting (e.g. if they forget to add it during upgrade).

File: classes/invitation/invitations/BaseInvitation.php [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+ */
+
+namespace PKP\invitation\invitations;
+
+use APP\core\Application;
+use Carbon\Carbon;
+use DateTime;
+use Illuminate\Queue\SerializesModels;
+use Illuminate\Support\Facades\Mail;
+use Illuminate\Mail\Mailable; // should be use PKP\mail\Mailable;
```
[the remark this answers is not in the document]
Comment (defstat) [reply classes/invitation/invitations/BaseInvitation.php:25]: This is more of a question. I think that `use PKP\mail\Mailable;` is needed to go along with the other mailables in the code base. I think templates are needed for that, though, and I don't have specific template for that. So I used the parent of `PKP\mail\Mailable` to send simple/sample emails with the accept/decline URLs. 

I changed it for `BaseInvitation`, but changes need to be done for the subclasses of that class when they will be developed.

File: classes/invitation/invitations/enums/InvitationType.php [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+ * @class InvitationStatus
+ *
+ * @ingroup invitations
+ *
+ * @brief Enumeration for invitation statuses
+ */
+
+namespace PKP\invitation\invitations\enums;
+
+class InvitationType
```
[the remark this answers is not in the document]
Comment (defstat) [reply classes/invitation/invitations/enums/InvitationType.php:19]: @asmecher I thought of that too. 

What do you say, to use the classname for the "type" purpose and completely remove the enumeration?

[the remark this answers is not in the document]
Comment (asmecher) [reply classes/invitation/invitations/enums/InvitationType.php:19]: Yes, agreed!

File: classes/invitation/invitations/BaseInvitation.php (hunk 12 above)
[the remark this answers is not in the document]
Comment (asmecher) [reply classes/invitation/invitations/BaseInvitation.php:25]: OK, it's fine to keep it as a Laravel mailable here, and subclasses can use specific mailables. :+1:

File: classes/invitation/models/Invitation.php [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+    {
+        $this->update([
+            'updated_at' => $this->currentTime(),
+            'status' => InvitationStatus::INVITATION_STATUS_CANCELLED
+        ]);
+    }
+
+    public function isExpired(): bool
+    {
+        $expiryDate = $this->expiry_date;
```
[the remark this answers is not in the document]
Comment (defstat) [reply classes/invitation/models/Invitation.php:181]: Fixed

File: classes/invitation/invitations/BaseInvitation.php [hunk 15]
```diff
@@ -0,0 +1,217 @@
...
+    {
+        $this->className = get_class($this);
+        $this->data = get_object_vars($this);
+    }
+
+    public function getPayload()
+    {
+        return [
+            "className" => $this->className,
+            "data" => $this->data
```
[the remark this answers is not in the document]
Comment (defstat) [reply classes/invitation/invitations/BaseInvitation.php:86]: Fixed

**Comment by defstat** [comment]
Closed in favor of #9200

**Pull request closed without merging**